fix(useMediaQuery): Return boolean describing media query match on first render

ArttuOll opened the pull request:
### What is the current behavior, and the steps to reproduce the issue?

Described in #1000 .

### What is the expected behavior?

The value returned from the hook should be a boolean describing the media query match on first render instead of `undefined`.

### How does this PR fix the problem?
By initializing the state with the the boolean received from `matchMedia` and creating a `MediaQueryList` entry to the cache in the process.

[Here](https://codesandbox.io/s/jovial-yalow-rt5mn7?file=/src/App.tsx) is a CodeSandbox, which contains the example provided in #1000, but with the new implementation of `useMediaQuery`. Note that the animation is no longer triggered when the page loads.

**This is a breaking change, since the value returned on the first render is now different than before. (Returning `undefined` before was not a bug since it was documented)**

## Checklist

- [X] Have you read [contribution guideline](../../CONTRIBUTING.md)?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [X] Is there an existing issue for this PR?
  - #1000 
- [X] Have the files been linted and formatted?
- [X] Have the docs been updated to match the changes in the PR?
- [X] Have the tests been updated to match the changes in the PR?
- [X] Have you run the tests locally to confirm they pass?

ArttuOll committed:
fix(useMediaQuery): Return boolean describing media query match on first render instead of undefined

This gets rid of the value changing from undefined to boolean on the first renders, which might
trigger undesired side-effects for users.

BREAKING CHANGE: On first render, boolean is returned instead of undefined.
fix #1000

xobotyi commented:
One problem - it will cause hydration error for SSR.

Initialisation should be flagged via hook parameters.

ArttuOll commented:
> One problem - it will cause hydration error for SSR.
> 
> Initialisation should be flagged via hook parameters.

Sounds doable. I'll implement this later.

ArttuOll commented:
Btw, could you elaborate on how this causes an hydration error? I wasn't aware of this.

xobotyi commented:
SSR will be first-rendered with undefined, client with some value - react will freak out on that.

xobotyi commented:
Also I have an idea to replace all that crap with `initializeWithValue` props in favor of context provider that will automatically tell all our hooks that we're in SSR mode without the need of passing extra arguments to hooks

ArttuOll commented:
> first render on server and in browser should match

Of course, makes sense! Thanks.

> Also I have an idea to replace all that crap with initializeWithValue props in favor of context provider that will automatically tell all our hooks that we're in SSR mode without the need of passing extra arguments to hooks

Interesting. Would the users then have to wrap their app in some sort of context provider?

xobotyi commented:
Yes, they should have, but in my eyes it is way more convenient that to be sure that you pass proper hook config to each hook that has isomorphic behavior.

ArttuOll committed:
feat(useMediaQuery): Add initializeWithValue option

This allows us to give the users the media query match state immediately on first render and
optionally give them SSR compatibility.

BREAKING CHANGE: Returned value is no longer undefined on first render. Set initializeWithValue to
false to re-enable this behavior.
fix #1000

ArttuOll commented:
@xobotyi How about something like I pushed in my last commit? I added `initializeWithValue` option and defaulted it to `true`, like with other hooks such as `useLocalStorageValue` etc.

xobotyi commented on src/useScreenOrientation/useScreenOrientation.ts:
```diff
@@ -9,7 +9,7 @@ export type ScreenOrientation = 'portrait' | 'landscape';
  */
 export function useScreenOrientation(): ScreenOrientation | undefined {
-  const matches = useMediaQuery('(orientation: portrait)');
+  const matches = useMediaQuery('(orientation: portrait)', { initializeWithValue: false });
```
Maybe we should also add an argument to this hook to yield correct value on init

ArttuOll replied:
Good point. Did not even think about this hook too much 😄 I'll work on this tomorrow.

ArttuOll replied:
Done.

ArttuOll committed:
Merge branch 'master' into pr/fix-1000

ArttuOll committed:
feat(useScreenOrientation): Initialize with correct value by default, optionally with undefined

This makes the hook work well with its underlying useMediaQuery hook. Also, many other hooks follow
this convention of initializing with correct value by default and optionally with undefined.

BREAKING CHANGE: Previously this hook initialized with undefined. Now it initializes by default with
the correct hook value. The previous behavior can be enabled by setting the initializeWithValue
option to false.

ArttuOll committed:
docs(useScreenOrientation): Update docs to describe the new options object

xobotyi commented:
:tada: This PR is included in version 20.0.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@react-hookz/web/v/20.0.0)
- [GitHub release](https://github.com/react-hookz/web/releases/tag/v20.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
